src/main/java/com/synopsys/integration/alert/web/security/authentication/AlertAuthenticationProvider.java

```diff
@@ -0,0 +1,115 @@
+        if (!supports(authentication.getClass())) {
+            throw new IllegalArgumentException("Only UsernamePasswordAuthenticationToken is supported, " + authentication.getClass() + " was attempted");
+        }
+        Authentication authenticationResult = performLdapAuthentication(authentication);
+        if (!authenticationResult.isAuthenticated()) {
+            authenticationResult = performDatabaseAuthentication(authentication);
+        }
+        final Collection<? extends GrantedAuthority> authorities = isAuthorized(authenticationResult) ? authenticationResult.getAuthorities() : List.of();
+        final UsernamePasswordAuthenticationToken authenticationToken = new UsernamePasswordAuthenticationToken(authenticationResult.getPrincipal(), authenticationResult.getCredentials(), authorities);
+        SecurityContextHolder.getContext().setAuthentication(authenticationToken);
+        return authenticationToken;
+    }
+
+    @Override
+    public boolean supports(final Class<?> authentication) {
+        return UsernamePasswordAuthenticationToken.class.isAssignableFrom(authentication);
+    }
+
+    private Authentication performDatabaseAuthentication(final Authentication pendingAuthentication) {
+        logger.info("Attempting database authentication...");
+        return alertDatabaseAuthProvider.authenticate(pendingAuthentication);
+    }
+
+    private Authentication performLdapAuthentication(final Authentication pendingAuthentication) {
+        logger.info("Checking ldap based authentication...");
+        Authentication result;
+        if (ldapManager.isLdapEnabled()) {
+            logger.info("LDAP authentication enabled");
+            try {
+                final LdapAuthenticationProvider authenticationProvider = ldapManager.getAuthenticationProvider();
+                result = authenticationProvider.authenticate(pendingAuthentication);
+            } catch (final AlertLDAPConfigurationException ex) {
+                logger.error("LDAP Configuration error", ex);
+                result = pendingAuthentication;
+            } catch (final Exception ex) {
+                logger.error("LDAP Authentication error", ex);
+                result = pendingAuthentication;
+            }
+        } else {
+            logger.info("LDAP authentication disabled");
+            result = pendingAuthentication;
+        }
+        return result;
+    }
+
+    private boolean isAuthorized(final Authentication authentication) {
+        final EnumSet<UserRole> allowedRoles = EnumSet.allOf(UserRole.class);
+        return authentication.getAuthorities().stream()
+                   .map(GrantedAuthority::getAuthority)
+                   .filter(role -> role.startsWith(UserModel.ROLE_PREFIX))
+                   .map(role -> StringUtils.substringAfter(role, UserModel.ROLE_PREFIX))
+                   .anyMatch(roleName -> allowedRoles.contains(UserRole.valueOf(roleName)));
```

Comment:
We COULD add one more map step here instead of calling the valueOf method inside. We would convert it to the following instead.
`.map(UserRole::valueOf)
.anyMatch(allowedRoles::contains)`

Reply:
sounds good I will make that change